File: Student_Model/concept.js
// Requires: Util/serialization.js, Util/zet.js
if (typeof window === "undefined") {
    var window = this;
}

Zet.declare('Concept', {
	superclass : Serialization.Serializable,
    defineBody : function(self){
		// Public Properties
		
		// Constructor
		self.construct = function construct(conceptId, name, description){
			self.inherited(construct);
			self._conceptId = conceptId;
            self._name = name;
            self._description = description;
		};

        self.eq = function(other){
            return (self.CLASS_ID === other.CLASS_ID &&
                self._conceptId === other._conceptId &&
                self._name === other._name &&
                self._description === other._description);
        };
        
        self.neq = function(other){
            return (!self.eq(other));
        };
		
		// Public Methods
		self.initializeFromToken = function initializeFromToken(token, context){
			self.inherited(initializeFromToken, [token, context]);
			self._conceptId = token.getitem('conceptId', true, null);
            self._name = token.getitem('name', true, null);
            self._description = token.getitem('description', true, null);
		};

		self.saveToToken = function saveToToken(){
            var token;
			token = self.inherited(saveToToken);
			token.setitem('conceptId', self._conceptId);
            token.setitem('name', self._name);
            token.setitem('description', self._description);
			return token;
		};
	}
});

File: Student_Model/Concept.py
# -*- coding: utf-8 -*-

from Util.Serialization import Serializable

class Concept(Serializable):
    """
    A concept related to some subject matter.
    """
    
    def __init__(self, conceptId=None, name=None, description=None):
        super(Concept, self).__init__()
        self._conceptId = conceptId
        self._name = name
        self._description = description

    def __eq__(self, other):
        return (type(self) == type(other) and
                self._conceptId == other._conceptId and
                self._name == other._name and
                self._description == other._description)

    def __ne__(self, other):
        return not self.__eq__(other)

    def getName(self):
        return self._name
    
    def getConceptId(self):
        return self._conceptId
    
    def getDescription(self):
        return self._description

    def saveToToken(self):
        token = super(Concept, self).saveToToken()
        token["conceptId"] = self._conceptId
        token["name"] = self._name
        token["description"] = self._description
        return token

    def initializeFromToken(self, token, context=None):
        super(Concept, self).initializeFromToken(token, context)
        self._conceptId = token["conceptId"]
        self._name = token["name"]
        self._description = token["description"]
